Add a filter to modify the sidebar

amieiro opened the pull request:
<!-- Thanks for contributing to GlotPress! Please, follow the GlotPress Contributing Guidelines:
https://github.com/GlotPress/GlotPress/blob/develop/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
This PR adds a filter to be able to modify the sidebar.

## Why?
<!-- Why is this PR necessary? What problem is it solving? What new functionality is it adding? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
We are adding some tabs in the sidebar [here](https://github.com/GlotPress/gp-translation-helpers/pull/148), so we need a filter to modify the sidebar content.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
This PR:
1. Adds a new filter. 
2. Modify the current info, storing the sidebar content into the `$meta_sidebar` variable, so it can be used in the filter. 
3. Add the new method `gp_link_user_get()` to create the link without printing it.

## Testing Instructions
<!-- Please include step-by-step instructions on how to test this PR. -->
<!-- 1. Open a original string in a project. -->
<!-- 2. Add the translation. -->
<!-- 3. etc. -->
1. Enable the [gp-translation-helpers](https://github.com/GlotPress/gp-translation-helpers) plugin, in the `tabs-sidebar` branch.
2. See the new content in the sidebar.

pedro-mendonca commented on gp-includes/template-links.php:
```diff
@@ -364,18 +364,33 @@ function gp_link_glossary_delete( $glossary, $set, $text = '', $attrs = array()
  * @param WP_User $user A WP_User user object.
  */
 function gp_link_user( $user ) {
+	// This content is scaped in the gp_link_user_get() method.
```
// This content is ~scaped~ escaped in the gp_link_user_get() method.
@amieiro here is a typo :)

amieiro replied:
Thank you! Solved.

akirk commented:
I was just thinking what people would need to do if they don't want the new sidebar. They could add a snippet like this:
```
add_filter(
    'wp_loaded',
    function() {
        remove_all_filters( 'gp_right_sidebar' );
    },
    99
);
```

akirk commented on gp-includes/template-links.php:
```diff
@@ -364,18 +364,33 @@ function gp_link_glossary_delete( $glossary, $set, $text = '', $attrs = array()
 function gp_link_user( $user ) {
+	// This content is escaped in the gp_link_user_get() method.
+	// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+	print( gp_link_user_get( $user ) );
```
We don't need to phpcs:ignore:
```suggestion
	echo wp_kses( gp_link_user_get( $user ), array( 'a' => array( 'href' => array(), 'tabindex' => array() ) );
```

amieiro replied:
It is ok, but it needs one more parenthesis at the end.

akirk commented on gp-includes/template-links.php:
```diff
@@ -364,18 +364,33 @@ function gp_link_glossary_delete( $glossary, $set, $text = '', $attrs = array()
+	$link = '';
 	if ( $user->display_name && $user->display_name !== $user->user_login ) {
-		printf(
+		$link = sprintf(
 			'<a href="%s" tabindex="-1">%s (%s)</a>',
 			esc_url( gp_url_profile( $user->user_nicename ) ),
 			esc_html( $user->display_name ),
 			esc_html( $user->user_login )
 		);
```
Could we simplify the logic here?

```suggestion
		return sprintf(
			'<a href="%s" tabindex="-1">%s (%s)</a>',
			esc_url( gp_url_profile( $user->user_nicename ) ),
			esc_html( $user->display_name ),
			esc_html( $user->user_login )
		);
```

akirk commented on gp-includes/template-links.php:
```diff
@@ -364,18 +364,33 @@ function gp_link_glossary_delete( $glossary, $set, $text = '', $attrs = array()
 			esc_html( $user->display_name ),
 			esc_html( $user->user_login )
 		);
 	} else {
```
```suggestion
	}
```

akirk commented on gp-includes/template-links.php:
```diff
@@ -364,18 +364,33 @@ function gp_link_glossary_delete( $glossary, $set, $text = '', $attrs = array()
 		);
 	} else {
-		printf(
+		$link = sprintf(
 			'<a href="%s" tabindex="-1">%s</a>',
 			esc_url( gp_url_profile( $user->user_nicename ) ),
 			esc_attr( $user->user_login )
 		);
 	}
+	return $link;
```
```suggestion
	return sprintf(
		'<a href="%s" tabindex="-1">%s</a>',
		esc_url( gp_url_profile( $user->user_nicename ) ),
		esc_attr( $user->user_login )
	);
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
-	<h3><?php _e( 'Meta', 'glotpress' ); ?></h3>
-
-	<?php gp_tmpl_load( 'translation-row-editor-meta-status', get_defined_vars() ); ?>
+$meta_sidebar  = '<div class="meta" id="sidebar-div-meta-' . $translation->row_id . '">';
```
```suggestion
$meta_sidebar  = '<div class="meta" id="sidebar-div-meta-' . esc_attr( $translation->row_id ) . '">';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
-
-	<?php gp_tmpl_load( 'translation-row-editor-meta-status', get_defined_vars() ); ?>
+$meta_sidebar  = '<div class="meta" id="sidebar-div-meta-' . $translation->row_id . '">';
+$meta_sidebar .= '<h3>' . __( 'Meta', 'glotpress' ) . '</h3>';
```
```suggestion
$meta_sidebar .= '<h3>' . esc_html__( 'Meta', 'glotpress' ) . '</h3>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+$meta_sidebar .= gp_tmpl_get_output( 'translation-row-editor-meta-status', get_defined_vars() );
+if ( $translation->context ) {
+	$meta_sidebar .= '		<dl>';
+	$meta_sidebar .= '			<dt>' . __( 'Context:', 'glotpress' ) . '</dt>';
```
```suggestion
	$meta_sidebar .= '			<dt>' . esc_html__( 'Context:', 'glotpress' ) . '</dt>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
-				<?php
+if ( $translation->extracted_comments ) {
+	$meta_sidebar .= '		<dl>';
+	$meta_sidebar .= '			<dt>' . __( 'Comment:', 'glotpress' ) . '</dt>';
```
```suggestion
	$meta_sidebar .= '			<dt>' . esc_html__( 'Comment:', 'glotpress' ) . '</dt>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+}
+if ( $translation->translation_added && '0000-00-00 00:00:00' !== $translation->translation_added ) {
+	$meta_sidebar .= '		<dl>';
+	$meta_sidebar .= '			<dt>' . __( 'Date added (GMT):', 'glotpress' ) . '</dt>';
```
```suggestion
	$meta_sidebar .= '			<dt>' . esc_html__( 'Date added (GMT):', 'glotpress' ) . '</dt>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+	$meta_sidebar .= '			<dd id="gmt-date-added-' . esc_attr( $translation->row_id ) . '">' . esc_html( $translation->translation_added ) . '</dd>';
+	$meta_sidebar .= '		</dl>';
+	$meta_sidebar .= '		<dl>';
+	$meta_sidebar .= '			<dt>' . __( 'Date added (local):', 'glotpress' ) . '</dt>';
```
```suggestion
	$meta_sidebar .= '			<dt>' . esc_html__( 'Date added (local):', 'glotpress' ) . '</dt>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+}
+if ( $translation->user ) {
+	$meta_sidebar .= '		<dl>';
+	$meta_sidebar .= '			<dt>' . __( 'Translated by:', 'glotpress' ) . '</dt>';
```
```suggestion
	$meta_sidebar .= '			<dt>' . esc_html__( 'Translated by:', 'glotpress' ) . '</dt>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+	$meta_sidebar .= '			<dt>';
 			if ( 'current' === $translation->translation_status ) {
-				_e( 'Approved by:', 'glotpress' );
+				$meta_sidebar .= __( 'Approved by:', 'glotpress' );
```
```suggestion
				$meta_sidebar .= esc_html__( 'Approved by:', 'glotpress' );
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+				$meta_sidebar .= __( 'Approved by:', 'glotpress' );
 			} elseif ( 'rejected' === $translation->translation_status ) {
-				_e( 'Rejected by:', 'glotpress' );
+				$meta_sidebar .= __( 'Rejected by:', 'glotpress' );
```
```suggestion
				$meta_sidebar .= esc_html__( 'Rejected by:', 'glotpress' );
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+$meta_sidebar .= ob_get_clean();
+
+$meta_sidebar .= '	<dl>';
+$meta_sidebar .= '		<dt>' . __( 'Priority:', 'glotpress' ) . '</dt>';
```
```suggestion
$meta_sidebar .= '		<dt>' . esc_html__( 'Priority:', 'glotpress' ) . '</dt>';
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+}
+ob_start();
+references( $project, $translation );
+$meta_sidebar .= ob_get_clean();
```
Can this be escaped with `wp_kses()`?

amieiro replied:
Done [here](https://github.com/GlotPress/GlotPress/pull/1531/commits/f4b49283fb5e75029bdcd74778a5b65ed2e918d1).

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
+				$meta_sidebar .= __( 'Rejected by:', 'glotpress' );
 			} else {
-				_e( 'Last updated by:', 'glotpress' );
+				$meta_sidebar .= __( 'Last updated by:', 'glotpress' );
```
```suggestion
				$meta_sidebar .= esc_html__( 'Last updated by:', 'glotpress' );
```

akirk commented on gp-templates/translation-row-editor-meta.php:
```diff
@@ -53,113 +53,118 @@
-	</dl>
-</div>
+// phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+echo $meta_sidebar;
```
Could we escape this output as well? The list of tags will be long but we'd be protected against mischief by the filter.

amieiro replied:
Done [here](https://github.com/GlotPress/GlotPress/pull/1531/commits/2ccb9cba87fc0929054b393c13e002324af9251c).

akirk reviewed (changes requested):
Not sure about the `bin/install-wp-tests.sh` file, was it added by mistake?

I wonder if we could keep the change smaller. We already need to resort to using `ob_start()` and `ob_get_contents()` for the `references()` call, couldn't we do it for the whole section? i.e. undo all of the changes and start with an `ob_start()` and end with a `$meta_sidebar = ob_get_clean();` and then pipe it into the filter.

ocean90 commented on gp-includes/template-links.php:
```diff
@@ -364,18 +364,37 @@ function gp_link_glossary_delete( $glossary, $set, $text = '', $attrs = array()
+	return sprintf(
+		'<a href="%s" tabindex="-1">%s</a>',
+		esc_url( gp_url_profile( $user->user_nicename ) ),
+		esc_attr( $user->user_login )
```
```suggestion
		esc_html( $user->user_login )
```

ocean90 reviewed (changes requested):
Can we please avoid adding unnecessary escaping unless it's actually input by a user? Translations are trusted so no need for `esc_html__()`, `wp_kses()` in `gp_link_user()` makes no sense since the data is already escaped in `gp_link_user_get()` and there's no reason to use `wp_kses()` for the whole sidebar since the sidebar markup is generated by us not by a user.

amieiro committed:
Add the filter with ob_start() and ob_get_clean()

Change the approach, with a simpler one, avoiding to update the HTML and
using the ob_start() and ob_get_clean() functions

amieiro commented:
@ocean90 @akirk I have updated the approach to add this filter, with a simpler one, avoiding to update the HTML and using the ob_start() and ob_get_clean() functions. Now the changes are minimal. I have used a "force push" to update the branch content.

ocean90 commented:
Have you already considered using the `gp_tmpl_load_locations` filter to change the path to a template which is in the plugin?

amieiro commented:
> Have you already considered using the `gp_tmpl_load_locations` filter to change the path to a template which is in the plugin?

Yes, I have considered this approach, but what we want to do, once this PR is merged, is to move the [gp-translation-helpers](https://github.com/GlotPress/gp-translation-helpers) plugin to the GlotPress core, so I prefer to filter the content and later to add the new tabs to GlotPress core.

ocean90 commented:
But there's now no difference between this filter and `gp_tmpl_load_locations`. Both can be used to replaced the content? I doubt that it's useful to extend the markup since it's just a bulk of HTML.


Tabs are also likely something we do not want to introduce in GlotPress as they provide a bad UX due to various factors like limited space and/or only having one panel visible.

amieiro commented:
@ocean90 you are right, I have created a new template with the tabs for the right sidebar in [this PR](https://github.com/GlotPress/gp-translation-helpers/pull/160), so I am going to close this PR. Thank you for the feedback.
